src/main/java/net/starlark/java/lib/toml/TomlParser.java

```diff
@@ -0,0 +1,567 @@
+          Object value = convertToJava(elem, semantics);
+          if (value != null) {
+            result.add(value);
+          }
+        } catch (EvalException ex) {
+          throw Starlark.errorf("at %s index %d: %s", Starlark.type(x), i, ex.getMessage());
+        }
+        i++;
+      }
+      return result;
+    }
+
+    // e.g. struct
+    if (x instanceof Structure || x instanceof NativeInfo) {
+      // Sort fields for determinism
+      List<String> fields =
+          Ordering.natural().sortedCopy(Starlark.dir(Mutability.IMMUTABLE, semantics, x));
+
+      Map<String, Object> result = new LinkedHashMap<>();
+      for (String field : fields) {
+        try {
+          Object v =
+              Starlark.getattr(
+                  Mutability.IMMUTABLE,
+                  semantics,
+                  x,
+                  field,
+                  null); // may fail (field not defined)
+          // Skip callables (methods)
+          if (x instanceof NativeInfo && v instanceof StarlarkCallable) {
+            continue;
+          }
+          Object value = convertToJava(v, semantics);
+          if (value != null) {
+            result.put(field, value);
+          }
+        } catch (EvalException ex) {
+          throw Starlark.errorf("in %s field .%s: %s", Starlark.type(x), field, ex.getMessage());
+        }
+      }
+      return result;
+    }
+
+    throw Starlark.errorf("cannot encode %s as TOML", Starlark.type(x));
+  }
+
+  /**
+   * Encodes a table (map) to TOML. In TOML, scalar values must come before nested tables
+   * to avoid them being associated with the wrong table header.
+   */
+  private static void encodeTable(StringBuilder sb, Map<String, Object> map, List<String> pathSegments) throws EvalException {
+    boolean needsBlankLine = false;
+
+    // First pass: output all scalar values and non-table arrays
+    for (Map.Entry<String, Object> e : map.entrySet()) {
+      Object value = e.getValue();
+      if (value instanceof Map) {
+        continue; // Handle tables in second pass
+      }
+      if (value instanceof List && isArrayOfTables((List<?>) value)) {
+        continue; // Handle array of tables in third pass
+      }
+      encodeKey(sb, e.getKey());
+      sb.append(" = ");
+      encodeValue(sb, value, 0);
+      sb.append("\n");
+      needsBlankLine = true;
+    }
+
+    // Second pass: output nested tables
+    for (Map.Entry<String, Object> e : map.entrySet()) {
+      Object value = e.getValue();
+      if (value instanceof Map) {
+        List<String> nestedPath = new ArrayList<>(pathSegments);
+        nestedPath.add(e.getKey());
+        if (needsBlankLine) {
+          sb.append("\n");
+        }
+        sb.append("[");
+        encodeTablePath(sb, nestedPath);
+        sb.append("]\n");
+        encodeTable(sb, (Map<String, Object>) value, nestedPath);
+        needsBlankLine = true;
+      }
+    }
+
+    // Third pass: output arrays of tables using [[array]] syntax
```

Comment:
You don't need a separate third pass for arrays of tables.

I believe this is valid TOML syntax:

```toml
[a]
lorem = "ipsum"

[a.b]
dolor = "sit"

[[a.c]]
amet = "consectetur"

[[a.c]]
adipiscing = "elit"

[a.d]
sed = "do"
```

It corresponds to
```json
{"a": {"lorem": "ipsum", "b": {"dolor": "sit"}, "c": [{"amet": "consectetur"}, {"adipiscing": "elit"}], "d": {"sed": "do"}}}
````